cylc/flow/data_store_mgr.py

```diff
@@ -766,100 +839,122 @@ def update_data_structure(self, updated_nodes=None):
-        # - Not in the set of pool points.
-        # - Not a source or target cycle point in the set of edges.
-        # This ensures a buffer of sources and targets in front and behind the
-        # task pool, while accommodating exceptions such as ICP dependencies.
-        # TODO: Turn nodes back to ghost if not in pool? (for suicide)
-        prune_points = set()
-        for s_point, t_points in list(self.edge_points.items()):
-            if (s_point not in self.pool_points and
-                    t_points.isdisjoint(self.pool_points)):
-                prune_points.add(str(s_point))
-                prune_points.update((str(t_p) for t_p in t_points))
-                del self.edge_points[s_point]
+
+        # Gather nodes to prune via a diff of nodes in the dependency paths
+        # of flagged boundary nodes against all other paths.
+        in_paths = [
+            v
+            for k, v in self.n_window_nodes.items()
+            if k in self.all_task_pool
+        ]
+        #    if k in self.all_task_pool
+        out_paths = [
+            v
+            for k, v in self.n_window_nodes.items()
+            if k in self.prune_flagged_nodes
+        ]
+        node_ids = set().union(*out_paths).difference(*in_paths)
+        # Absolute triggers may be present in task pool, so recheck.
+        # Clear the rest.
+        self.prune_flagged_nodes.intersection_update(self.all_task_pool)
+
+        tp_data = self.data[self.workflow_id][TASK_PROXIES]
+        tp_added = self.added[TASK_PROXIES]
+        parent_ids = set()
+        for tp_id in node_ids:
+            if tp_id in self.n_window_nodes:
+                del self.n_window_nodes[tp_id]
+            if tp_id in self.n_window_edges:
+                del self.n_window_edges[tp_id]
+            if tp_id in tp_data:
+                node = tp_data[tp_id]
+            elif tp_id in tp_added:
+                node = tp_added[tp_id]
+            else:
                 continue
-            t_diffs = t_points.difference(self.pool_points)
-            if t_diffs:
-                prune_points.update((str(t_p) for t_p in t_diffs))
-                self.edge_points[s_point].difference_update(t_diffs)
-        # Action pruning if any eligible cycle points are found.
-        if prune_points:
-            self.prune_points(prune_points)
-        if new_points or prune_points:
-            # Pruned and/or additional elements require
-            # state/status recalculation, and ID ref updates.
+            self.deltas[TASK_PROXIES].pruned.append(tp_id)
+            self.schd.job_pool.remove_task_jobs(tp_id)
+            self.deltas[EDGES].pruned.extend(node.edges)
+            parent_ids.add(node.first_parent)
+
+        prune_ids = set()
+        checked_ids = set()
+        while parent_ids:
+            for fp_id in parent_ids:
+                self._family_ascent_point_prune(
+                    fp_id, node_ids, parent_ids, checked_ids, prune_ids)
+                break
```

Comment:
Could we remove the `for` loop with

```py
while parent_ids:
    self._family_ascent_point_prune(next(iter(parent_ids)), node_ids, parent_ids, checked_ids, prune_ids)
```

I **think** that'd be equivalent to iterating once with `for` and `break`'ing right away? (haven't tested it, sorry)

Reply:
That would be equivalent to iterating until `parent_ids` are empty .. Will have to look at it again, but I think (like the family update code), I was trying an alternative to `.pop()` where you remove the item at the same time.